src/System.Buffers/src/System/Buffers/DefaultArrayPool.cs

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace System.Buffers
+{
+    internal sealed class DefaultArrayPool<T> : ArrayPool<T>
+    {
+        private const int MinimiumArraySize = 16;
+        private int _maxArrayLength;
+        private int _numberOfArraysPerBucket;
+        private bool _isTObjectType;    // cache this to only do the reflection call once
+        private ArrayPoolBucket<T>[] _buckets;
+        private int _rentedBufferCount;
+        private int _totalBufferCount;
+
+        public override int RentedBufferCount
+        {
+            get
+            {
+                return Volatile.Read(ref _rentedBufferCount);
+            }
+        }
+
+        public override int TotalBufferCount
+        {
+            get
+            {
+                return Volatile.Read(ref _totalBufferCount);
+            }
+        }
+
+        internal DefaultArrayPool(int maxLength, int arraysPerBucket)
+        {
+            // Our bucketing algorithm has a minimum length of 16
+            if (maxLength < MinimiumArraySize)
+                throw new ArgumentOutOfRangeException(nameof(maxLength), SR.Maximum_Array_Size_Under_Minimum_Supported_Value);
+            if (maxLength <= 0)
+                throw new ArgumentOutOfRangeException(nameof(maxLength));
+            if (arraysPerBucket <= 0)
+                throw new ArgumentOutOfRangeException(nameof(arraysPerBucket));
+
+            int maxBuckets = Utilities.SelectBucketIndex(maxLength);
+            _isTObjectType = typeof(T).GetTypeInfo().IsClass;
+            _maxArrayLength = maxLength;
+            _numberOfArraysPerBucket = arraysPerBucket;
+            _buckets = new ArrayPoolBucket<T>[maxBuckets + 1];
+            _rentedBufferCount = 0;
+            _totalBufferCount = 0;
+        }
+
+        public override T[] Rent(int length)
+        {
+            if (length <= 0)
+                throw new ArgumentOutOfRangeException(nameof(length));
+
+            Interlocked.Increment(ref _rentedBufferCount);
+
+            int index = Utilities.SelectBucketIndex(length);
+            if (index >= _buckets.Length)
+            {
+                Interlocked.Increment(ref _totalBufferCount);
+                return new T[length];
+            }
+            else
+            {
+                T[] buffer = null;
+
+                // Search for an array starting at the 'index' bucket. If the bucket
+                // is empty, bump up to the next higher bucket and try that one
```

Comment:
We should think through this heuristic carefully.  Let's say there are two main uses of the pool: requesting a 1K buffer and requesting a 1MB buffer.  And let's say that someone puts back a 1MB buffer and there are no other buffers in the pool. Then someone comes along and requests a 1K buffer, and is given the 1MB buffer.  Then someone comes along and requests a 1MB buffer, at which point we need to allocate 1MB when we could have just allocated 1K.  Obviously there are also cases where this heuristic would provide great results; I'm simply pointing out that without exploring it on concrete expected use cases, we could actually be making things worse.  Just something to keep in mind.